#166840878 User signup error swag

tolumide-ng opened the pull request:
#### What does this PR do?
- Enables user to signup on errorswag

#### Description of Task to be completed?
- User can signup on errorswag

#### How should this be manually tested?
- Navigate to the signup page on errorswag

#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?(if applicable)
#166840878

#### Screenshots
[166840878](https://www.pivotaltracker.com/story/show/166840878)

Onyimatics commented on src/pages/Signup/index.js:
```diff
@@ -0,0 +1,117 @@
+    <div className="w-7/12 p-10">
+      {/* logo and inputs */}
+      <div className="w-full flex flex-col items-center p-2 mx-auto">
+        {/* <div className="w-5/6 mx-auto bg-white mt-10 mb-4 text-center flex flex-col items-center"> */}
```
you may need to remove these comments please.

Onyimatics commented on src/fakeData.js:
```diff
@@ -0,0 +1,40 @@
+      }
+    }
+  ]
+};
```
Please what is the value of this file to this PR?

Onyimatics reviewed (changes requested):
Good job @tolumide-ng and @olorunwalawrence.
Please endeavour to remove all the unneeded comments on your code.

segunolalive commented on tests/signup.spec.js:
```diff
@@ -0,0 +1,84 @@
+});
+const fake = jest.fn();
+
+describe('it should work', () => {
```
This should be the component being described

segunolalive commented on tests/signup.spec.js:
```diff
@@ -0,0 +1,84 @@
+  it('renders the component', () => {
+    expect(wrapper.exists()).toBe(true);
+    expect(wrapper.find('Form').exists()).toBe(true);
+    expect(wrapper.find('button').exists()).toBe(true);
```
I think asserting that there's a submit button (type=submit)/ the text in the button is more useful than just that there's a button.

Assert the number of buttons. Other important form elements as well.

segunolalive replied:
This test doesn't actually test that the component signs a user up. And that is reasonable because the component doesn't sign a user up. The action/network request does that. The component only displays what is appropriate based on the application state.

Also, add test for negative paths e.g when a user enters invalid data.

**Hypothetical Solution**
```js
const  jest.fn(() => 'lol')

state =  { signUp }

// then in your test.

sbmitBtn.simulate('click');

// then you should be able to do:

expect(signUp).toHaveBeenCalled();
expect(signUp).toHaveBeenCalledTimes(1);

```

segunolalive reviewed (changes requested):
@olorunwalawrence @tolumide-ng 

This is good. Tests need a bit more work but good job guys.

codexempire commented on src/pages/Signup/index.js:
```diff
@@ -0,0 +1,142 @@
+import { Formik, Form, Field } from 'formik';
+import { signupAction } from '../../store/modules/auth/actions/signup';
+import logo from '../../assets/images/logo.png';
+// import signupCard from '../../assets/images/signup.png';
```
Remove comments from the code

codexempire commented on src/pages/Signup/index.js:
```diff
@@ -0,0 +1,142 @@
+
+import SignupSchema from './schema';
+
+export const SignupForm = ({ isLoading, error, status, signUp }) => (
```
Notice you exported the function regularly and also exported the same function as a default would love to know what it's for, but I do think it won't be necessary to export a function regularly and also as the default

tolumide-ng replied:
Thank you for asking. Please check the test module here for clarification. 
https://github.com/andela/ah-kifaru-frontend/blob/ft-user-signup-errorSwag-166840878/src/pages/Signup/spec.js

codexempire commented on src/components/App.js:
```diff
@@ -1,12 +1,19 @@
-      <Routes />
-    </Router>
+    <Provider store={store}>
+      <Router>
```
Can you move the `<Router></Router>` wrapping the `<Routes />` into the route folder so as to group similar things together

tolumide-ng replied:
This would not be moved, for the following reasons. 
We need a component called getUser which is already been built by @kechyy at this point to restore the user when a refresh occurs. Also we need to separate concerns across the application to improve a single functionality.

timi-codes replied:
As @codexempire mentioned, the TTL suggested these changes since the App and Router are simple components. Kindly revert back to what we have as @kechyy also have this changes.

codexempire reviewed (commented):
Nice work Noticed the integration of various packages including **`toastr`** left some comment behind

Apart from the comment I also think you should try removing **`.DS_STORE`** from the **`PR`**

timi-codes reviewed (changes requested):
Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.

You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
and this
https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

segunolalive commented on src/store/modules/auth/reducer/index.js:
```diff
@@ -0,0 +1,17 @@
+      return { ...state, ...payload };
+    default:
+      return state;
+  }
```
You could have done this since you've done the heavy lifting in your action creators.
```js
function authReducer(state = auth, { type, payload }) {
  return authActions.includes(type) ? { ...state, ...payload } : state;
}

```

codexempire commented on src/store/modules/auth/actions/index.js:
```diff
@@ -0,0 +1,58 @@
+    });
+    const { token } = response.data.data;
+    saveToLocalStorage({ token, url });
+    const user = decodeToken({ history });
```
The after signup the user should be told to verify their accounts and the page may not have to be directed to another. Rather it might just display a message showing successful registration so the user could verify his/her account

tolumide-ng replied:
As much as I agree that a user should be notified to check their email address for the verification link (which can be accomplished via Toastr). I disagree with not redirecting them to the page they were initially on before signing up or the homepage of the application incase they were previously not on an errorSwag page that needed user authentication. Not only does this improve the user experience, it also makes the application more accessible

More so, I think we already discussed this during our state management meeting for authentication last week.

codexempire reviewed (changes requested):
Nice work please attend to the comment

tolumide-ng commented:
> Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
> You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.
> 
> You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
> and this
> https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

Thank you for your observation. The tests are already been catered for by @kechyy since we are using the same actions and reducers

Onyimatics commented on src/components/SocialMedia/social.spec.js:
```diff
@@ -77,7 +79,7 @@ describe('SocialMediaLogin', () => {
     );
 
-    expect(localStorage.getItem('token')).toEqual(userToken);
+    // expect(localStorage.getItem('token')).toEqual(userToken);
```
I suggest you reach out next time, instead of commenting this `assertion` out.
You may need to replace that assertion with 
```js
expect(JSON.parse(localStorage.getItem('token'))).toEqual(userToken);
Thank you.
```

tolumide-ng replied:
> I suggest you reach out next time, instead of commenting this `assertion` out.
> You may need to replace that assertion with
> 
> ```js
> expect(JSON.parse(localStorage.getItem('token'))).toEqual(userToken);
> Thank you.
> ```

Hello @Onyimatics , thank you for the review. However, there is no problem with the previous assertion. I only forgot to remove the comments after testing what I needed on this part of the code. Thank you.

Onyimatics commented on src/pages/Signup/index.spec.js:
```diff
@@ -0,0 +1,96 @@
+  location,
+  status: 'authenticationLoading',
+  SignUp: jest.fn(({ userData = {}, history, location }, cb) => {
+    cb({ data: true });
```
I suggest you consider making the name of this `cb` more descriptive.

tolumide-ng replied:
Okay

Onyimatics commented on src/pages/Signup/index.spec.js:
```diff
@@ -0,0 +1,96 @@
+  );
+};
+
+describe.only('SignupForm User Input Validation', () => {
```
I guess you must have forgotten to remove this `.only`

tolumide-ng replied:
Yh, I did. Thank you

Onyimatics commented on src/store/modules/auth/actions/index.js:
```diff
@@ -33,24 +34,35 @@ export const authFailure = error => ({
-    const { message } = error.response.data;
+    return url ? history.push(url) : history.push('/articles');
+  } catch ({ response }) {
+    const { message } = response.data;
```
instead of removing these lines 
```js
    } catch (error) {
    const { message } = error.response.data;
```, I suggest you use an `if` statement there to also cater for the error coming from `error.response.data` , incase a user encounters a `network error`.

tolumide-ng replied:
Oh True. However, In the case of a network failure, `axios` still returns a response since that it is what was stated on the try block. 
Hence, I think 
``` const message = response && response.data && response.data.message ```
might be okay. What do you think?

Onyimatics reviewed (changes requested):
Well done guys. You may need to check out the comments I left.

tohbay commented on __test__/App.spec.js:
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { BrowserRouter, Switch } from 'react-router-dom';
 import thunk from 'redux-thunk';
+import { mount } from 'enzyme';
```
Hi @tolumide-ng and @olorunwalawrence,

Please remove the `mount` function since it has been configured globally on the project.

tolumide-ng replied:
Okay

tohbay commented on __test__/components/navbar.spec.js:
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { Provider } from 'react-redux';
 import configureStore from 'redux-mock-store';
+import { mount } from 'enzyme';
```
Hi @tolumide-ng and @olorunwalawrence,

Please remove the `mount` function since it has been configured globally on the project.

olorunwalawrence replied:
sorry this mount is on navbar component not on the signup

tohbay commented on enzyme.config.js:
```diff
@@ -0,0 +1,4 @@
+import { configure } from 'enzyme';
```
I think there is no need for this file `enzyme.config.js`, the configurations therein are already implemented on `__test__/setup.js`

tolumide-ng replied:
Okay

tohbay reviewed (changes requested):
Good job @tolumide-ng  and @olorunwalawrence 👍 

There are 4 commits, please squash them. Also, attend to the comments.

tolumide-ng commented:
> Good job @tolumide-ng and @olorunwalawrence 👍
> 
> There are 4 commits, please squash them. Also, attend to the comments.

Thank you for your review. However, commits cannot be squashed yet as instructed by the TTL since this is a collaborated PR

timi-codes commented on package.json:
```diff
@@ -92,38 +93,47 @@
+    "formik": "^1.5.8",
     "glob": "^7.1.4",
     "html-webpack-plugin": "^3.2.0",
+    "i": "^0.3.6",
```
kindly remove this as it must have been installed in err.

timi-codes commented on package.json:
```diff
@@ -92,38 +93,47 @@
     "jest-svg-transformer": "^1.0.0",
     "jwt-decode": "^2.2.0",
     "mini-css-extract-plugin": "^0.8.0",
+    "npm": "^6.11.0",
```
kindly remove this as it must have been installed in err.

tolumide-ng replied:
👍

tolumide-ng commented:
> Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
> You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.
> 
> You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
> and this
> https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

Thank you @timi-codes . The actions, thunk and reducers are not tested at all in this PR since this feature was implemented in conjunction with the login feature (assigned to @tobechukwuobitube and @kechyy ), they are already testing the actions and reducers appropriately. Thank you once again

timi-codes commented on src/index.css:
```diff
@@ -0,0 +1,9 @@
+@tailwind base;
```
Please use the `main.scss`for global stylings

tolumide-ng replied:
I guess, this was when you changed the file configuration to sass as it was initially index.css when I setup the tailwindcss
I would remove that now

timi-codes reviewed (changes requested):
@olorunwalawrence @tolumide-ng Good job on this pull request. However, Kindly attend to all the changes as soon as possible so that we can have this merged today.

tolumide-ng commented:
> @olorunwalawrence @tolumide-ng Good job on this pull request. However, Kindly attend to all the changes as soon as possible so that we can have this merged today.

Changes attended. Please rereview

tolumide-ng committed:
feat(users): User signup on errorSwag

    - Corrects bug on password non-conformity
    - Adds Animation to show loading during user signup
    - Adds Social Media Buttons for user signup authentication
    - Fixes CSS on user signup
    - Implements state and reducers for signup
    - Creates UI for errorswag signup page
    - Returns relevant error message on user signup action
    - Implements recommended feedback on UI label and redux state params
    - Implements state and reducers for signup
    - Creates UI for errorswag signup page
    - Returns relevant error message on user signup action
    - Implements recommended feedback on UI label and redux state params

[Delivers #166840878]

segunolalive commented on src/pages/Signup/index.js:
```diff
@@ -0,0 +1,48 @@
+  <div className="w-11/12 md:w-9/12 lg:w-7/12 mx-auto m-16 rounded-lg flex items-stretch bg-red-400 h-100">
+    <div className="md:flex max-w-lg w-7/12 bg-color height items-center signup-card hidden rounded-l-lg ">
+      <div className=" pl-8 font-bold text-left text-white text-2xl">
+        <div>Welcome back,</div>
```
Change `Welcome back,` to `Welcome to ErrorSwag`.

Wrap this text in an `h1` tag or the most appropriate heading tag. This is important for accessibility and SEO.

Use `<br>` tags to break the text into multiple lines if needed.

segunolalive reviewed (changes requested):
🎊  This is good.

Make the correction on the SignUp form header and merge.
